Source/Core/DiscIO/VolumeVerifier.cpp

```diff
@@ -0,0 +1,648 @@
+{
+constexpr u64 MINI_DVD_SIZE = 1459978240;  // GameCube
+constexpr u64 SL_DVD_SIZE = 4699979776;    // Wii retail
+constexpr u64 SL_DVD_R_SIZE = 4707319808;  // Wii RVT-R
+constexpr u64 DL_DVD_SIZE = 8511160320;    // Wii retail
+constexpr u64 DL_DVD_R_SIZE = 8543666176;  // Wii RVT-R
+
+constexpr int BLOCK_SIZE = 0x8000;
+
+VolumeVerifier::VolumeVerifier(const Volume& volume)
+    : m_volume(volume), m_started(false), m_done(false), m_progress(0),
+      m_max_progress(Common::AlignUp(volume.GetSize(), BLOCK_SIZE) / BLOCK_SIZE)
+{
+}
+
+void VolumeVerifier::Start()
+{
+  ASSERT(!m_started);
+  m_started = true;
+
+  m_is_tgc = m_volume.GetBlobType() == BlobType::TGC;
+  m_is_datel = !GetBootDOLOffset(m_volume, m_volume.GetGamePartition()).has_value();
+  m_is_not_retail =
+      (m_volume.GetVolumeType() == Platform::WiiDisc && !m_volume.IsEncryptedAndHashed()) ||
+      IsDebugSigned();
+
+  const std::vector<Partition> partitions = CheckPartitions();
+  if (m_volume.GetVolumeType() == Platform::WiiWAD)
+    CheckCorrectlySigned(PARTITION_NONE, GetStringT("This title is not correctly signed."));
+  CheckDiscSize();
+  CheckMisc();
+}
+
+std::vector<Partition> VolumeVerifier::CheckPartitions()
+{
+  const std::vector<Partition> partitions = m_volume.GetPartitions();
+  if (partitions.empty())
+  {
+    if (m_volume.GetVolumeType() != Platform::WiiWAD &&
+        !m_volume.GetFileSystem(m_volume.GetGamePartition()))
+    {
+      AddProblem(Severity::High, GetStringT("The filesystem is invalid or could not be read."));
+      return {};
+    }
+    return {m_volume.GetGamePartition()};
+  }
+
+  std::optional<u32> partitions_in_first_table = m_volume.ReadSwapped<u32>(0x40000, PARTITION_NONE);
+  if (partitions_in_first_table && *partitions_in_first_table > 8)
+  {
+    // Not sure if 8 actually is the limit, but there certainly aren't any discs
+    // released that have as many partitions as 8 in the first partition table.
+    // The only game that has that many partitions in total is Super Smash Bros. Brawl,
+    // and that game places all partitions other than UPDATE and DATA in the second table.
+    AddProblem(Severity::Low,
+               GetStringT("There are too many partitions in the first partition table."));
+  }
+
+  std::vector<u32> types;
+  for (const Partition& partition : partitions)
+  {
+    const std::optional<u32> type = m_volume.GetPartitionType(partition);
+    if (type)
+      types.emplace_back(*type);
+  }
+
+  if (std::find(types.cbegin(), types.cend(), PARTITION_UPDATE) == types.cend())
+    AddProblem(Severity::Low, GetStringT("The update partition is missing."));
+
+  if (std::find(types.cbegin(), types.cend(), PARTITION_DATA) == types.cend())
+    AddProblem(Severity::High, GetStringT("The data partition is missing."));
+
+  const bool has_channel_partition =
+      std::find(types.cbegin(), types.cend(), PARTITION_CHANNEL) != types.cend();
+  if (ShouldHaveChannelPartition() && has_channel_partition)
```

Comment:
I'm guessing this should probably be `if (ShouldHaveChannelPartition() && !has_channel_partition)`?